modules/data.remote/R/GEDI_AGB_prep.R

```diff
@@ -0,0 +1,527 @@
+#' @importFrom rlang .data
+GEDI_AGB_extract <- function(site_info, start_date, end_date, outdir, nfile.min = 0, nrow.min = 0, buffer = 0.01) {
+  #Initialize the multicore computation.
+  if (future::supportsMulticore()) {
+    future::plan(future::multicore)
+  } else {
+    future::plan(future::multisession)
+  }
+  # grab site.info and convert from lat/lon to sf objects of points and buffer areas.
+  GEDI_AGB <- split(as.data.frame(site_info), seq(nrow(as.data.frame(site_info)))) %>%
+    furrr::future_map(function(point){
+      # flag determine if we have satisfied res.filter object.
+      csv.valid <- FALSE
+      # extends for filter.
+      extends <- c(point$lat + buffer,
+                   point$lat - buffer,
+                   point$lon - buffer,
+                   point$lon + buffer) %>% 
+        purrr::set_names(c("ymax", "ymin", "xmin", "xmax"))
+      # redirect to the current folder.
+      # if we already create the folder.
+      site_folder <- file.path(outdir, point$site_id)
+      if (file.exists(site_folder)) {
+        # if csv file has been generated.
+        if (length(list.files(site_folder, "GEDI_AGB.csv", full.names = T)) > 0) {
+          # read csv file.
+          res <- utils::read.csv(list.files(site_folder, "GEDI_AGB.csv", full.names = T))
+          if (file.exists(file.path(site_folder, "extends.txt")) & nfile.min != 0) {
+            extends <- utils::read.table(file.path(site_folder, "extends.txt"), skip = 1, col.names = c("ymax", "ymin", "xmin", "xmax"))
+            extends <- extends[nrow(extends),]
+          }
+          # filter previous records based on space and time.
+          res.filter <- res %>% dplyr::filter(.data$lat_lowestmode <= extends["ymax"],
+                                              .data$lat_lowestmode >= extends["ymin"],
+                                              .data$lon_lowestmode >= extends["xmin"], 
+                                              .data$lon_lowestmode <= extends["xmax"],
+                                              lubridate::as_date(.data$date) >= lubridate::as_date(start_date),
+                                              lubridate::as_date(.data$date) <= lubridate::as_date(end_date))
+          # determine if res.filter is not empty.
+          if (nrow(res.filter) > 0) {
+            csv.valid <- TRUE
+          }
+        }
+      } else {
+        # create folder for the current site.
+        base::dir.create(site_folder)
+        # copy and paste credentials to the current folder.
+        base::file.copy(file.path(outdir, "netrc"), file.path(site_folder, "netrc"))
+      }
+      # filter out point outside the GEDI spatial domain.
+      if (point$lat > 52 | point$lat < -52) {
+        utils::write.table("Point is outside the GEDI domain.", file = file.path(site_folder, "Error.txt"))
+        return(NA)
+      }
+      # if we have previous GEDI records covering space and time.
+      if (csv.valid) {
+        return(res.filter)
```

Comment:
I think there is an assumption that if we have the observations (currently by determining if the `nrow` of valid records is greater than 0) that cover `space` and `time`, we will use the observation instead of re-downloading it. Upon your question, I think in the future, we will need to document file name of each pre-downloaded `GEDI L4A` files such that any new files within the range will be downloaded and processed.

Reply:
Just documented that need as comments in the script.

Reply:
Might be worth opening an Issue about this so that the need to extend this function in the future isn't solely left to your 🧠 and future users deciding to take a close look at the code internals.